### ISSUE TITLE
Consider restoring DT[col] api

### ISSUE THREAD

**Issue by st-pasha** [h2oai/datatable#1910]
In #1395 this API was deprecated and removed. However, I'm wondering whether it was a good judgment. In fact, we have 3 choices for `DT[a]`: treat `a` as a column selector ("option 1"), treat it as a row selector ("option 2"), or make it an error ("option 3"). Currently this form raises an exception (so option 3), but the message says it will be treated as row-selector in 0.9 (option 2).

The best way to think about this problem is in terms of examples, so let's start:

- `DT[0]`: this could mean either the first row, or the first column. Generally, I think, selecting columns is much more useful (and faster) than selecting individual rows, giving more weight to option 1. However when the index value is large `DT[1000000]`, people tend to think that it refers to the row number. Obviously, we can't make the interpretation dependent on the magnitude of the value. 

- When the integer index is in a variable with a neutral name: `DT[i]`, the expression more naturally looks like a column selector than a row selector.

- Since iteration over a Frame produces columns, it seems natural for `DT[i]` to mean the same as `list(DT)[i]`, meaning option 1.

- `DT[:10]` -- does this mean the first 10 rows, or the first 10 columns? This better be consistent with the meaning of an integer selector. However, in this case the row-interpretation (option 2) comes to mind easier, especially when the numbers are large: `DT[:1000000]` looks like the first 1M rows rather than columns. 

- `DT["id"]` -- under option 1 this selects column with the name "id", whereas under option 2 this is not valid (unless maybe if DT is keyed by a string column, in which case this could mean selecting a row with index "id"). This form strongly favors option 1, especially since this syntax is natural to people who have experience with pandas.

- `DT[:]` means "select everything" both under option 1 and option 2, so this is unambiguous.

- `DT[None]` could also mean "select everything", or it could be an error, or it could mean "select nothing and return None". The latter option may be useful when trying to select a particular column (or row?) dynamically, and in some cases you want to not select anything.

- `DT[f.A > 0]` is not valid under option 1, but becomes a row filter under option 2. Obviously, under option 1 one could write `DT[f.A>0, :]`.

- Allowing column selector `DT["id"]` reinforces the view of a Frame as an ordered dictionary of columns; in particular it allows syntax `dict(DT)` and `**DT` (similar to `*DT` which is already allowed). In particular, cbinding two frames could be written as `dt.Frame(**DT1, **DT2)`.

In summary, it seems to me that interpreting a single-element selector as _column_ is more useful than interpreting it as a _row_. 

Thoughts / feedback welcome.

**Comment by oleksiyskononenko** [comment]
I foresee the following problem with the column selector approach. Consider one does ‘DT[i]’ selecting i’th column, but later wants to select j’th row in i’th column by naturally adding ‘j’ to the end of expression as ‘DT[i, j]’. Is there any way we can prevent such a counterintuitive datatable behavior if we go for a column selector approach?

**Comment by st-pasha** [comment]
Yes, it is somewhat counterintuitive that `DT[i, j]` shortens to `DT[j]` instead of `DT[i]`. It is not completely unheard-of, though, for example `range(a, b)` shortens to `range(b)`. In our case I think the interest of being able to fit into python's APIs is more important than having consistent argument order...

To answer your question, if both `i` and `j` are integers, then there is no way for us to know whether the user means (row, col) or (col, row). The potential for confusion exists even without the proposed changes, though they'll probably exacerbate the problem. I think even more dangerous could be the pattern `DT[i][j]`. I guess, they'll have to learn these things the hard way...

We could also look at other APIs for comparison:

## R data.table 

R data.table follows the <option 2> approach: `DT[i,j] -> DT[i,] -> DT[i]`. In particular, `DT[1]` selects the first row and `DT['A']` doesn't work. However, in R there are alternative quick ways to select a single column: `DT[[1]]` or `DT$A`, neither of which are available in python. In R `[[i]]` and `$A` are considered standard ways of accessing elements of a list / dictionary, meaning R data.table can implement those and still fit into R's API, while being free to interpret `[i]` however it wants.

## R data.frame

R data.frame follows the <option 1> approach: both `DF[1]` and `DF['A']` return the first column 'A'. 

## pandas 

For a pandas dataframe:
- `df[0]` is an error;
- `df[:5]` returns the first 5 rows;
- `df['A']` returns the column 'A';
- `df[df['A'] > 0]` returns rows filtered by the expression "A>0";
- `list(df)` returns the list of column names
- `list(reversed(df))` is an error;
- `dict(df)` returns a dictionary {column name: Series};
- `**df` works the same as `**dict(df)`.

**Comment by goldentom42** [comment]
Hi @st-pasha and @oleksiyskononenko,

Just would like to add a few thoughts on this.

The way pandas makes the difference between columns and rows indexing uses a `.loc` or `.iloc`, which I don't think is something we want to do. So to get special rows using integers you would do : `df.iloc[row_indices]` and here you can't provide column names unless you do an ugly `df[col_list].iloc[row_raw_indices]`. 
If you want to select row by index values then you can go with `df.loc[row_index, col_list]`. 

And then the difference between `.iloc` and `.loc` is that `iloc` works on the actual row number while `loc` works on the index number if an index is set, which in turn is very confusing.

Even if pandas has this col selector I believe we want to stick to the H2O.Frame way of doing things, which uses `df[:, col_list]`. 

The idea behind that is to help people who already love our open source solution go to `datatable` :) 

And again thanks for all your hard work !

**Comment by st-pasha** [comment]
Thanks Olivier,

Just to clarify: this proposal does not affect the standard `DT[i, j]` syntax at all. It merely adds a new interpretation of what it means to say `DT[i]`: this selects a single column. Selecting one column is actually a very common operation, so it's nice to have a more succinct syntax for it. Other than niceties , though, is that it will make `Frame` better satisfy the requirements of python "mapping" interface, allowing us to use a frame object whenever a mapping is expected.

**Comment by goldentom42** [comment]
Thx for the clarification @st-pasha !
